scheduler/data-flow/src/main/kotlin/io/seldon/dataflow/kafka/Pipeline.kt

```diff
@@ -51,31 +89,38 @@ class Pipeline(
         }
         streams.setStateListener(this)
         streams.start()
+        status = PipelineStatus.StreamStarting()
 
         // Do not allow pipeline to be marked as ready until it has successfully rebalanced.
         latch.await()
 
-        return !errorOrShutdown
+        return status
     }
 
     fun stop() {
+        // defend against stop() being called while start() is still waiting on the latch
+        latch.countDown()
+
+        // close needs to be called even if streams.start() never gets called
         streams.close()
         // Does not clean up everything see https://issues.apache.org/jira/browse/KAFKA-13787
         streams.cleanUp()
+        status = PipelineStatus.StreamStopped()
     }
 
-    override fun onChange(newState: KafkaStreams.State?, oldState: KafkaStreams.State?) {
+    override fun onChange(newState: State?, oldState: State?) {
         logger.info { "pipeline ${metadata.name} (v${metadata.version}) changing to state $newState" }
-        if (newState == KafkaStreams.State.RUNNING) {
+        if (newState == State.RUNNING) {
+            status = PipelineStatus.Started()
             latch.countDown()
             return
         }
         // CREATED, REBALANCING and RUNNING (with the latter one already handled above)
         // are the only non-error states. Everything else indicates an error or shutdown
         // and we should release the lock on which start() awaits and return an error.
         // see: https://kafka.apache.org/28/javadoc/org/apache/kafka/streams/KafkaStreams.State.html
-        if (newState != KafkaStreams.State.CREATED && newState != KafkaStreams.State.REBALANCING) {
-            errorOrShutdown = true
+        if (newState != State.CREATED && newState != State.REBALANCING) {
+            status = PipelineStatus.Error(newState).withMessage("pipeline data streams error: kafka streams state: $newState")
```

Comment:
**Review note:**

At the moment, the status updates happening in the kafka streams state update callback are not used anywhere. However, this lays the groundwork for being able to access them via the Pipeline.state property.

We'll need to be careful in selecting which of those changes should be propagated to the scheduler, in order to let it take global decisions about the fate of the pipeline (i.e re-init, stop across other components like {pipeline, model}gateway, etc)